### ISSUE TITLE
Read-Host : Add Trim Option

### ISSUE THREAD

**Issue by aksarben** [PowerShell/PowerShell#19680]
### Summary of the new feature / enhancement

Can you add an optional parameter to **Read-Host** to automatically remove leading & trailing whitespace (**string.trim()**)?

Sometimes users accidentally add white space that can cause string comparisons to fail. Since whitespace is invisible, it sometimes can take a few minutes to figure out why the failure is happening.

**Comment by mklement0** [comment]
I agree that this would be helpful, but just to note the _workaround_: `$userInput = (Read-Host 'Enter something').Trim()`

**Comment by aksarben** [comment]
Yup, that workaround does the trick. But as with all workarounds, of course, it's not part of the official documentation and many people won't think of it.

___________________________

The Cyber Hymnal™ <><
http://www.hymntime.com/tch
________________________________

**Comment by rhubarb-geek-nz** [comment]
> it's not part of the official documentation and many people won't think of it.

I don't count it as a workaround at all, it is simply using PowerShell

```
PS> help Read-Host
..snip..
OUTPUTS
    System.String
    System.Security.SecureString
```

I suggest that it is documented.

[String.Trim](https://learn.microsoft.com/en-us/dotnet/api/system.string.trim?view=net-7.0)

**Comment by mklement0** [comment]
@rhubarb-geek-nz, from that perspective _nothing_ that can be implemented in PowerShell code - no matter how cumbersome and non-obvious - constitutes a workaround.

The point here is that wanting to trim user input is the _typical_ use case, yet:

* `Read-Host` doesn't even support this functionality as an _opt-in_ - implementing one is what this proposal is about.

* Absent direct support, the next best thing is to have the documentation _directly_ address this typical use case, by spelling out the `(Read-Host ...).Trim()` implementation - and not making users _infer_ this implementation via multiple levels of indirection.
   * I've created https://github.com/MicrosoftDocs/PowerShell-Docs/issues/10088 to suggest adding a concrete example to the `Read-Host` help topic.

**Comment by jhoneill** [comment]
> It would be silly if every command that returned a string then has to have all the additional options to uppercase it , lowercase it, trim it, pad it and split it etc.

Very much this.

see https://github.com/PowerShell/PowerShell/issues/16660#issuecomment-999110578  for why I think the vast majority of the time `read-host` should not be used at all.  It's correct/incorrect usage ratio is worse than `write-host`

> The point here is that wanting to trim user input is the _typical_ use case

I think that about 99% of the time Read-Host is used because the author is writing like 1970s basic and asking the user to type things in when they should, in fact, be asking for parameters.

e.g.
```
param (
 [parameter(HelpMessage='Enter the X dimension of the thing we're calculating' , Mandatory=$true)]
$x
)
```
not 
```
$x = Read-host 'Enter the X dimension of the thing we're calculating' 
```

How often in practice does the user type spaces which should not form part of the input?  we're probably more concerned that user doesn't give empty input 
so we do
```
do {$x = Read-host "Please input X"} until ($x -ne "")
```
but no one is asking for an option for that, nor to ensure that input is numeric nor split comma separated input into multiple items or any of the many other things which might be bolted on `Read-Host` and benefit someone. 

Adding `-MethodProvidedByReturnType`  is generally **not** done. The usual thing is to do `(command).MethodProvidedByReturnType(xx)`  or even `Read-Host "Enter x" | % trim` (discouraged for scripts)

We might well say `Get-Content` should have `-Trim` or `-Split` or `-Replace` and they'd be more useful than in Read-Host, but we don't because we expect people to be able to do those operations on strings and demand it of things which get strings.

**Comment by rhubarb-geek-nz** [comment]
Just like [Go To Statement Considered Harmful](https://homepages.cwi.nl/~storm/teaching/reader/Dijkstra68.pdf) the same could be said of [Read-Host](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/read-host).

Any script with Read-Host is much less reusable by another script and it can't be used in a job. I agree with the sentiment that params should be used instead, you can even use the params to read secure strings.

```
#!/usr/bin/env pwsh
param([parameter(Mandatory=$true)][System.Security.SecureString]$password)

$UnsecurePassword = (New-Object PSCredential 0, $password).GetNetworkCredential().Password

$UnsecurePassword
```

**Comment by mklement0** [comment]
`Read-Host` and `Write-Host` are essential for _creating terminal-based user interfaces_, primarily - but not exclusively - for non-technical end users (who may not know or care that PowerShell is being used to implement the script).

This is completely unrelated to what you're talking about (parameters for technical users familiar with PowerShell), and this conflation - especially if used to argue against `Read-Host` and `Write-Host` _as commands_ - is unhelpful.

The argument re not providing a switch for something that can also be done manually,  after the fact, is specious:
While separation of concerns is important, there are many examples of _convenience parameters_ that reduce the verbosity and repetitiveness of code, and possibly help performance; case in point: 

Why have `Get-Content -First 10 ...` when you can do `Get-Content ... | Select-Object -First 10`? Because it is more concise, more convenient (and performs better, though that's not the most important aspect).

**Comment by jhoneill** [comment]
@mklement0  I wonder if you are just playing devil's advocate,  

There are a few cases where we display a menu or ask the user for a selection, or pause until the user hits enter which need `Read-Host`. But when someone not very experienced with PowerShell uses `read-host` it's usually because they're writing to that 1970s BASIC paradigm of getting users to type parameters after the program starts and having it produce formatted strings written to the console. We need `Write-Host` to separate _output_ from _stuff for people to look at_, which is a distinction those programs don't make. The commands are needed but also facilitate worst practice - imagine if the _only_ way you could use `man` was to run `man` with no parameters and then answer a prompt  for what you wanted help on - that would be a dreadful bit of coding.

Where the dividing line lies between "use two cmdlets"  like `gci | sort-object length` and "we'll give it to the user in one as a convenience" like `get-content -first` seems to be what seemed like a good idea at the time to the original designers.  Generally the principles of    
(1) move selection of what goes down the pipeline as far to the left as possible  and then    
(2) don't do in the cmdlet something that the returned objects can do     
seem to be fairly consistent.  Generally "command  -only-return-the-first n"  is better than "command | stop-command-after n"

**Comment by rhubarb-geek-nz** [comment]
> Generally "command -only-return-the-first n" is better than "command | stop-command-after n"

Even that may not be true. Consider a directory listing where technically there is no order to the files in a directory, they are just a set of files. If you had a Get-ChildItem stop after n you could get a different set each time. If you let the Cmdlet on the left focus on the getting of the data but the cmdlets to the right do the filtering and sorting then you make the cmdlets that get the data much simpler, and end up with more consistent sorting for data from different sources.

It is worth comparing with SQL, where selecting rows just gives you the rows and they could be in any order, it is not until you do WHERE and ORDER BY and LIMIT that then you get the filtering and restriction in number of rows. Of course the difference with a database query engine is that it constructs a plan across the whole statement in order to be the most efficient. With cmdlets we don't have that luxury. For instance we have no mechanism for a Cmdlet on the right to tell a Cmdlet on the left that it has provided enough pipeline elements, the Cmdlet on the left will just keep on going.

I am more in favour of keeping the data providers as simple as possible, and then make the sorting and filtering cmdlets the most powerful and flexible because both the reuse and consistency will be so much better,

**Comment by mklement0** [comment]
> The commands are needed but also facilitate worst practice

This is a completely separate debate - bringing it up here is inappropriate and only causes confusion.

**Comment by jhoneill** [comment]
> > Generally "command -only-return-the-first n" is better than "command | stop-command-after n"
> 
> Even that may not be true. Consider a directory listing where technically there is no order to the files in a directory, they are just a set of files. 

I only know for sure with FAT, but I think it is true of all file systems, the directory keeps the files in creation order. I think .NET (and possibly the windows API) sorts before returning.   But I don't think the API has a "Ask the file system for 10 files in this directory", we get all the files and if we only want 10 the work has been done. 

> If you had a Get-ChildItem stop after n you could get a different set each time. If you let the Cmdlet on the left focus on the getting of the data but the cmdlets to the right do the filtering and sorting then you make the cmdlets that get the data much simpler, and end up with more consistent sorting for data from different sources.

Yes that is generally true.  But if I can call a rest api which returns 1,000,000 items and apply a where on the client side, or I can call the API with the where condition and return 1 item. It's plainly better not to bring the other 999,999 over the wire and send them down the pipeline. If I have a function which calls an API which usually returns about 100 items I'll probably sort by name rather than tell users to do it, because 9 times out of 10 they want that order, and for the 10th the cost of the extra sort doesn't matter. 

> 
> It is worth comparing with SQL, where selecting rows just gives you the rows and they could be in any order, it is not until you do WHERE and ORDER BY and LIMIT that then you get the filtering and restriction in number of rows. Of course the difference with a database query engine is that it constructs a plan across the whole statement in order to be the most efficient. With cmdlets we don't have that luxury. For instance we have no mechanism for a Cmdlet on the right to tell a Cmdlet on the left that it has provided enough pipeline elements, the Cmdlet on the left will just keep on going.
>
Actually there is a mechanism and Select-Object uses it.  try  ` dir / -rec -ea 0 | Select -first 2 ` and ` dir / -rec -ea 0 | out-null`   

AIUI it raises a special terminating error, which is not available to other things. If it is a the SQL query or rest API that gets a ton  of stuff, it only stops that going down the pipeline but for something like `get-content` on a million line file, it does stop the rest being  read   So `gc foo | select -first 10` is better than `(gc foo)[0..9]`  but `Get-Sql $sql | select -first 10` going to get all the data and we'd be better ensuring $sql had "top 10"   
 
> I am more in favour of keeping the data providers as simple as possible, and then make the sorting and filtering cmdlets the most powerful and flexible because both the reuse and consistency will be so much better,

Not over-complicating is good :-)  If you think of `get-process` the API lets us get processes by name or ID, so it would be silly to get all processes and apply a `where-object`.  But there is no need to have `get-process` able to select or sort based on memory used, that's why commands are composed from cmdlets in the first place.  


 



> > The commands are needed but also facilitate worst practice
> 
> This is a completely separate debate - bringing it up here is inappropriate and only causes confusion.

If anyone wants to know why I'm not taking this to the cmdlet working group, and why if another member does I'd argue it's not even something we should accept a community PR on, then I've laid out why. Sorry for confusing you.

**Comment by mklement0** [comment]
> Sorry for confusing you.

Glad to hear it. I encourage you and @rhubarb-geek-nz to hide all your comment here as "Off Topic."

**Comment by rhubarb-geek-nz** [comment]
>  it's not part of the official documentation and many people won't think of it.

While it is part of the official documentation;

PS> help Read-Host
..snip..
OUTPUTS
    System.String
    System.Security.SecureString
 
And [System.String.Trim](https://learn.microsoft.com/en-us/dotnet/api/system.string.trim) it may do a disservice to the very audience you are trying to help.

If people think they are writing pure, portable PowerShell script may find either

```
PS> Read-Host -Trim   
-Trim: 
```

or

```
PS> Read-Host -Prompt "hello" -Trim
Read-Host: A parameter cannot be found that matches parameter name 'Trim'.
```

If they run their script on a previous version of PowerShell.

So they now need to be aware that their script will only run on PowerShell 7.4.0 or with PSReadLine greater than 2.2.6

Then they will need a "using module PSReadLine" with a ModuleVersion, I could not find the syntax for that.

If they publish their script in `PSGallery`, likewise they will need a `PowerShellVersion` field in their module manifest.

So unless [Publish-Module](https://learn.microsoft.com/en-us/powershell/module/powershellget/publish-module) parses the scripts for unsupported parameters you may end up with published modules, while tested and working with one version fail on previous versions of PowerShell Core or on WindowsPowerShell.

**Comment by mklement0** [comment]
@rhubarb-geek-nz, since the off-topic ship has sailed, allow me to summarize your arguments - I do hope I got the gist right:

* "Typical use cases aren't worth documenting (with examples), because what you need to implement them can be gleaned, piece by piece, from other documentation."

* "New features aren't worth introducing, because old language versions won't have them."

**Comment by rhubarb-geek-nz** [comment]
Not quite

- What was called a workaround isn't really a workaround, being able to use dotnet directly is fundamental part of how PowerShell works
- New features can certainly be added but be aware of the backward compatibility issues that may arise.

**Comment by mklement0** [comment]
* _Workaround_  is a relative term that pertains to the inability to implement a task as intended, due to a (perceived) limitation or a bug: that limitation / bug must then be _worked around_ to achieve the desired result. 
  * That's how I used the term: relative to the perceived limitation of `Read-Host` not offering the desired trimming itself.
  * This terminology aspect has nothing to do with the merits of providing an example implementing a _typical use case_ in the `Read-Host` help topic. Such an example - albeit a more elaborate one than originally proposed - has since been added: https://github.com/MicrosoftDocs/PowerShell-Docs/pull/10089

* Of course backward compatibility issues can arise, but that is no reason to halt the progress of the language.
  *  Such issues apply to _any_ new feature, and are therefore not worth discussing in the context of a _particular_ feature request.
  *  A feature request should be discussed _on its own merits_ and - if deemed a worthwhile addition in principle - only _then_, potentially, a discussion of _when to introduce it_ can be had (e.g., bundled with similar new features, hold until the next major release, ...)

**Comment by jhoneill** [comment]
> > Sorry for confusing you.
> 
> Glad to hear it. I encourage you and @rhubarb-geek-nz to hide all your comments here as "Off Topic." (I'm happy to hide my responses thereafter.)

If I see additional people who are confused I'll be sure to do that. 

> * _Workaround_  is a relative term that pertains to the inability to implement a task as intended, due to a (perceived) limitation or a bug: that limitation / bug must then be _worked around_ to achieve the desired result.

While true, this does not really help matters. `Read-Host` does not convert to upper/lower, split, validate, trim, or do other replacements on the string it returns nor (AFAIK) do other analogs of the BASIC'S `INPUT` command (which I think mirrors one in Fortran). These are considered normal things to do to a string once it has been fetched. Someone thinking that it would those string operations were merged into commands (and why stop at `Read-Host`) doesn't mean the normal way of doing things is a problem to be worked around.  

> * A feature request should be discussed _on its own merits_ and - if deemed a worthwhile addition in principle - only _then_

As the only person from the cmdlets wg who has given this any time I haven't seen anything to persuade me - after an extensive discussion - that it is worth discussing in the wg. @aksarben  can we close this now ?

**Comment by mklement0** [comment]
> If I see additional people who are confused I'll be sure to do that.

Given how this thread has unfolded, I'm sure those who stumble upon this will be - you may simply never hear from them.
Not hearing from others is no justification to let off-topic comments stand.

> As the only person from the cmdlets wg who has given this any time 
>  @aksarben can we close this now ?

It is inappropriate to speculate about the intent of other workgroup members (without having received actual feedback).
It is inappropriate to unilaterally lobby for closure of a feature request.

----

> `Read-Host` does not convert to upper/lower, split, validate, trim, or do other replacements 

_That_ - unlike many other comments in this thread - is an _on-topic_ argument; it has been addressed before, but let me try to clarify:

* No one is asking for _arbitrary_ string transformations / operations to be baked into `Read-Host`, though _common_ ones are worth considering, such as `-Trim`, if they're relevant to _interactive prompting_:

* Users, especially non-technical ones, accidentally typing spaces before or after the input is a real-world pitfall that anyone writing _robust_ user interfaces must guard against (this doesn't preclude having to perform _additional_ operations on the string returned).
   * Ideally, the logic would be _reversed_: `Read-Host` should trim leading and trailing whitespace _by default_, unless explicitly asked not to; that ship has sailed for reasons of backward compatibility.
   * `$userInput = Read-Host -Trim  'Enter something'` is the next best thing.

**Comment by rhubarb-geek-nz** [comment]
> * No one is asking for _arbitrary_ string transformations / operations to be baked into `Read-Host`, though _common_ ones are worth considering, such as `-Trim`, if they're relevant to _interactive prompting_:

Me! Me! I am. The comparison with param was made and how it will translate to given type and also provide autocompletion. It would be really useful for interactive applications to have the same capability, eg at a top level menu to use Read-Host to return an enum value of the valid options, likewise with numbers only accept numerical input, ideally with ranges, or values from a ValidateSet.
Similarly a consistent way of aborting a current data entry session to return to the main menu by using control-C.

**Comment by santisq** [comment]
There is nothing stopping you from replacing `Read-Host` with your own wrapper function that handles trimming, it could even be in your `$profile`. I think developers time would much better invested in other issues.

```powershell
function Read-Host {
    [CmdletBinding(
        DefaultParameterSetName = 'AsString',
        HelpUri = 'https://go.microsoft.com/fwlink/?LinkID=2096610'
    )]
    param(
        [Parameter(Position = 0, ValueFromRemainingArguments = $true)]
        [AllowNull()]
        [System.Object]
        ${Prompt},

        [Parameter(ParameterSetName = 'AsSecureString')]
        [switch]
        ${AsSecureString},

        [Parameter(ParameterSetName = 'AsString')]
        [switch]
        ${MaskInput},

        [Parameter(ParameterSetName = 'AsString')]
        [switch]
        ${Trim}
    )

    begin {
        $withTrim = $PSBoundParameters.Remove('Trim')
        $scriptCmd = { Microsoft.PowerShell.Utility\Read-Host @PSBoundParameters }
        $steppablePipeline = $scriptCmd.GetSteppablePipeline($myInvocation.CommandOrigin)
        $steppablePipeline.Begin($true)
    }

    process {
        if($withTrim) {
            return ([string] $steppablePipeline.Process()).Trim()
        }
        $steppablePipeline.Process()
    }

    end {
        $steppablePipeline.End()
    }

    clean {
        if ($null -ne $steppablePipeline) {
            $steppablePipeline.Clean()
        }
    }
}
````

**Comment by mklement0** [comment]
Thanks for the code, @santisq  - I encourage you to publish it as a self-answered question on Stack Overflow to give it wider exposure, along with a link to this issue: that may draw more users who'd be interested in seeing this implemented as part of `Read-Host` itself (and therefore thumbs-ups, which are a gauge of community interest).

Anyone who'd put such a proxy function in their `$PROFILE` would clearly prefer that this functionality _be part of the language_ (the built-in commands). And anyone wanting to write reusable scripts relying on it would have to bundle the proxy function with their code, which is obviously a pain.

> I think developers time would much better invested in other issues.

Given how trivial this would be to implement, I think that's a false dichotomy - there is no either-or here.
More time was wasted in off-topic discussions here than it probably would have taken to implement the feature.

Again, the only question should be: is this a worthwhile addition to `Read-Host`?
If deemed so, prioritization of developer time may _then_ come into play, reflected in when the green-lit enhancement is actually implemented.

Whether a given feature is worthwhile is - at least initially, pending a potential appeal, as codified [here](https://github.com/PowerShell/PowerShell/blob/master/docs/community/working-group.md) - is to be decided by the relevant _working group_ (and how much time passes before that happens is also a reflection of priorities, I suppose).

It certainly shouldn't be decided unilaterally by a single member of a working group, especially in an underhanded manner (removal of the "Needs-Triage" tag, withholding of the relevant "WG-*" tag, badgering the OP to close their issue).

**Comment by doctordns** [comment]
Just a small point. I have always taught that **All user is input is dangerous and evil until you prove otherwise (and even then...).**.
Speaking personally, I have learned long ago that in a few cases, trailing (or leading) spaces can be problematic, and quite happily use `.trim()` in the PowerShell code.  

I have not seen a huge demand for this kind of feature either. I may be looking in the wrong place! ;-)

I learned long ago that there is no such thing as "trivial to implement". And the implantation is not the only cost involved. Anything that unnecessarily adds to the PowerShell package is to be avoided. I am not overly.  

I will add this issue to the Cmdlet WG's agenda although I can't promise how quickly we can get to it.

**Comment by mklement0** [comment]
@doctordns

> I will add this issue to the Cmdlet WG's agenda although I can't promise how quickly we can get to it.

Thank  you.
That is all I was asking for - please don't forget to tag the issue accordingly.

(I won't defend the merits of this proposal further - I think all angles have been covered.)

**Comment by doctordns** [comment]
@mklement0 thanks for the reminder - I have labeled this issue accordingly.

**Comment by jhoneill** [comment]
@mklement0  I'm going to disengage from your comments  , as it appears others have had the wisdom to do before me.  

The post no longer need triaging. I removed that label. I did not close it because (a) the OP might want to keep it open, (b) other members of the WG might have a different view to mine as @doctordns  has now shown.  It is very clear from everything I have said where my opinion as a one member of that WG, ends and where other members need to speak for themselves.  But my opinion (which will now go to WG) is as in the 5 points above. I will also have a word with the powers that be about the definition of trolling and possible ways forward.

**Comment by doctordns** [comment]
@mklement0 As it happens< i too am not convinced this is needed. So it;s not quite a unilateral decision. I am happy to have the WG discuss it, but I am not in favour.

**Comment by Andrew74L** [comment]
> Yup, that workaround does the trick. But as with all workarounds, of course, it's not part of the official documentation and many people won't think of it.

What about a string editing cmdlet...?
```
$userInput = Read-Host 'Enter something'
Edit-String -String $userInput -Trim [-Case Upper|Lower|Sentence] [-Join <string>] [-Prefix <string>] [-Suffix <string>]
```
Example use:
```
PS > $fullname = Read-Host 'Enter your full name'
Enter your full name: First Last
PS > Edit-String -String $fullname -Trim -Case Lower -Join '.'
PS > Edit-String -String $fullname -Join '@' -Suffix 'somewhere.com'
PS > $fullname
first.last@somewhere.com
```

**Comment by doctordns** [comment]
A string editing cmdlet might be useful to some users. I do not think there is enough support for putting such a command into the core of PowerShell. 

For me, the tools for editing a string that exist now are the System.String class's methods, Regular expressions and the PowerShell string operators (eg `-Split`). These work for me in the few cases where I need to play with such conversions. 

Perhaps consider creating a function and publishing it in a Community module on the PS Gallery.

**Comment by jhoneill** [comment]
> For me, the tools for editing a string that exist now are the System.String class's methods, Regular expressions and the PowerShell string operators (eg `-Split`). These work for me in the few cases where I need to play with such conversions.

We have -Join -split -replace  (and those of us who think in regexes can use -replace to do a trim operation) but no `-trim` operator. I don't think there is a need for a `-ToUpper` or `-ToLower`  but I can see others might. Which string functions are good for operators and which aren't isn't clear cut. 

@Andrew74L  It might make a good community function, but 
 `Edit-String -String $fullname -Join '@' -Suffix 'somewhere.com'`  is easier as  `$fullname += '@somewhere.com'` 
and 
`Edit-String -String $fullname -Trim -Case Lower -Join '.'` is easier as `$fullname = $fullname.trim().ToLower.Replace(' ','.')`

**Comment by santisq** [comment]
@jhoneill there is no need for a `-Trim` or `-ToUpper` or `-ToLower` really when `ForEach-Object` can invoke the member for you:

```powershell
' foo ' | ForEach-Object Trim
```

I don't see the point in adding more layers of abstraction when they're not needed. This is preventing folks from learning .NET in my honest opinion.

**Comment by jhoneill** [comment]
> @jhoneill there is no need for a `-Trim` or `-ToUpper` or `-ToLower` really when `ForEach-Object` can invoke the member for you:
> 
> ```powershell
> ' foo ' | ForEach-Object Trim
> ```
> 
> I don't see the point in adding more layers of abstraction when they're not needed. This is preventing folks from learning .NET in my honest opinion.

Fair point, although  `( 'foo ').trim()`  where anything can go in the first '()' is just easier.  
Arguably, there's no need for `-Split`  but it's convenient.  (and `'foo.bar'.Split('.')`  is not the same thing as `'foo.bar' -Split '.'`). If someone postpones learning about string objects and their methods because `-Split` exists they probably don't need that knowledge yet. Which different from making every string method a function so that they can postpone indefinitely.

**Comment by santisq** [comment]
@jhoneill for reference, `-split` is `regex.Split` not `string.Split` but yes overall I'm agreeing with you.

**Comment by Andrew74L** [comment]
> A string editing cmdlet might be useful to some users. I do not think there is enough support for putting such a command into the core of PowerShell.
> 
> For me, the tools for editing a string that exist now are the System.String class's methods, Regular expressions and the PowerShell string operators (eg -Split). These work for me in the few cases where I need to play with such conversions.

Might an intermediate level user feel more comfortable using a dedicated cmdlet?

> Perhaps consider creating a function and publishing it in a Community module on the PS Gallery.

Okay

> Such a command would not, however, resolve directly this issue.

Maybe not, but it would sidestep the debate over convenience parameters, and once the command exists, parameters could be added (e.g., `-reverse`) without causing compatibility issues.

> It's on the agenda for our next Cmdlet WG meeting although we may not get to it (we only have an hour!).

Okay, great.

**Comment by Andrew74L** [comment]
> @Andrew74L It might make a good community function, but `Edit-String -String $fullname -Join '@' -Suffix 'somewhere.com'` is easier as `$fullname += '@somewhere.com'` and `Edit-String -String $fullname -Trim -Case Lower -Join '.'` is easier as `$fullname = $fullname.trim().ToLower.Replace(' ','.')`

Agree for the first example. Disagree for the second - the hypothetical command is more readable and at least as fast to type.

**Comment by jhoneill** [comment]
> @jhoneill for reference, `-split` is `regex.Split` not `string.Split`  

I have been aware of that for a decent length of time :-)    `-replace` and `-split` are case-insensitive  regex, and the string methods are case sensitive. I hit a problem with someone's build script which used the `.Replace()` because something in VSCode was changing the case so the replacement didn't happen and the build exploded.  

> > @Andrew74L It might make a good community function, but `Edit-String -String $fullname -Join '@' -Suffix 'somewhere.com'` is easier as `$fullname += '@somewhere.com'` and `Edit-String -String $fullname -Trim -Case Lower -Join '.'` is easier as `$fullname = $fullname.trim().ToLower.Replace(' ','.')`
> 
> Agree for the first example. Disagree for the second - the hypothetical command is more readable and at least as fast to type.


The mooted function seemed to split on space and then re-join which would be odd behaviour.  As shown it changes the string passed in (less usual but not unprecedented - `Add-Member` does that) , but in some cases the original needs to be preserved, so if I have     
`$displayName = "James O'Neill" `    
and I want to make an email address from it, I want `$email = _something with $displayname_`    
and then I can `new-user $displayname   $email` 

So often that second example would need at least `$email = $displayname` first. A lot of the time it would be a mix of existing operators and string methods.  But a command would be useful to some people some of the time.  So I doubt it would get included in PowerShell itself. A full spec in separate issue would be a better way to get it looked at.

**Comment by SteveL-MSFT** [comment]
@PowerShell/wg-powershell-cmdlets reviewed this proposal and rejects the proposed addition to `Read-Host`.   Users are expected to validate any external input before use which includes, but not limited to, trimming whitespace.  Since we do not believe this scenario is sufficiently common, it does not make sense to add general string methods to this cmdlet.  It may make sense to have an updated example in the docs for validating input which includes trimming whitespace.

**Comment by mklement0** [comment]
@SteveL-MSFT , re documentation:

* The `Read-Host` topic now has [Example 4](https://learn.microsoft.com/en-us/powershell/module/Microsoft.PowerShell.Utility/Read-Host?view=powershell-7.3#example-4-normalizing-input) which shows the use of `.Trim()` (although the example is a bit complex).

* The original issue I had opened in order to get an example added - https://github.com/MicrosoftDocs/PowerShell-Docs/issues/10088 - has been reopened as a reminder that a general string-manipulation conceptual topic is apparently forthcoming.

**Comment by mklement0** [comment]
@SteveL-MSFT:

 I assume you've come across my complaint about inappropriate unilateral behavior in this thread, summarized [here](https://github.com/PowerShell/PowerShell/issues/19680#issuecomment-1561639220), so there's need for me to report it separately anymore. 

  * However, a general complaint procedure, as suggested in #19730 (which I now feel comfortable linking to, given the comment there discussed below), is still worth implementing.

I think said behavior is at odds with the mandates of working-group members, as already implied by the existing documents describing the WG duties.

However, one additional inappropriate behavior that unfolded in the wake of this thread strikes me as worth adding as a DO NOT to [Working Groups (WGs)](https://github.com/PowerShell/PowerShell/blob/master/docs/community/governance.md#working-groups-wgs):

 * It is inappropriate for WG members to _block_ community members, as it robs them of the ability to respond / react to those members' posts / comments and - last but not least - also from being _notified_ of @-mentions in comments by said members.
 
 * Case in point: Due to being blocked by @jhoneill, I was _not_ notified of his comment at https://github.com/PowerShell/PowerShell/issues/19730#issuecomment-1580672493, which explicitly @-mentions me - and made the issue about a _general_ complaint mechanism _personal_. I discovered it only _accidentally_, and had no one else ever responded there again, I could have missed it altogether.
 
---

To address said comment:

* It is a problematic for a WG member not to be able to tell the difference between arguments and personal attacks.

* It is even more problematic to then make accusations of such attacks, threaten the other with the Code of Conduct, and generally project their own disruptive behavior onto others.

* There is clearly no awareness of the inappropriateness of the unilateral actions, as the following sentence attests:
  *  > I only take things to the WG which I think are worth the time of its members.

 * @SteveL-MSFT, if - to my surprise - you think it _is_ appropriate for a single WG member to decide what does and doesn't _even jointly get triaged as a feature request_, i.e. to act as the _sole gatekeeper_, do let us know.
   * If so - which, needless to say, I oppose - perhaps that should be added to the WG document, for future clarity.

**Comment by aksarben** [comment]
Thanks for your consideration…

___________________________

The Cyber Hymnal™ <><
http://www.hymntime.com/tch
________________________________

**Comment by mklement0** [comment]
@jhoneill, by and large I appreciate the measured response and the  accountability regarding the label, but let me address some important points:

On the meta aspects:

> What we have had instead has wasted many hours

Fundamentally, if a better way forward emerges from this, the effort wasn't wasted.
The effort certainly could have been reduced, however, if you had shown accountability earlier, instead of ignoring my arguments and blocking me.

> if I make code-of-conduct complaints

Frankly, the only code-of-conduct complaint that would have been appropriate in the context of this thread is against yourself, for unwarranted threats with the code of conduct and accusations of trolling.

It is indeed your choice whether to respond, though doing so _in good faith_ is always preferable, as long as there actual differences to resolve (as opposed to something having devolved into bickering).

---

I can fully appreciate the need to weed out issues ahead of time, so they don't require discussion in a quorum, which includes:
* Closing as a  duplicate.
* Closing as an answered question.
* Closure by the user themselves, _of their own volition_,  _after a respectful and constructive discussion_.

Clearly, the latter has _not_ happened here.

> The default assumption for any request to provide access to a return-object's method(s) via one or more cmdlet-parameters would be "no"

I've addressed this before, but let me try again:

* The framing isn't correct: it's not about _access to methods_ it's about _adding functionality that is a sensible part of the cmdlet's purpose_ - that it involves something that can be _also_ be done later and that the proposed parameter name matches the name of a method on the object returned is _incidental_.

* Clearly, such "convenience duplication" of functionality  is both _common_ in the existing cmdlets and, if I were to guess, _frequently  used_ (certainly by me).

  * It gave us such things as  `Get-Content -First`, `Sort-Object -Top` (a fairly recent addition), `Select-String -Raw`, and there are probably many more examples.

It is precisely because what are to me misreadings of a feature request can occur that it is important not to operate on such general rules, especially not unilaterally.

---

To recap: The larger issue here is not that you expressed an _opinion on the feature request at hand_, but that:

* You deliberately withheld the `WG-*` label.
* You removed the `Needs-Triage` label.

The combination of these two actions _in effect_ would have made this feature request die on the vine, with you as the sole arbiter.

Even if the removal of `Needs-Triage` was an honest mistake (and it sounds like that was the case) - and clarifying the proper use of this label is what made me open #19731 (and I'm glad to hear that you now want clarity there too) -  the withholding of the `WG-*` label _alone_ is problematic, especially based on how you framed this decision:

> If anyone wants to know why I'm not taking this to the cmdlet working group

Is there any other way to read this than that you're thinking of yourself as the gatekeeper of what does and doesn't get discussed by the quorum of members?

Sure, there's no obligation for any single WG member to be in charge of assigning the `WG-*` label.
In fact, the [working-group document](https://github.com/PowerShell/PowerShell/blob/master/docs/community/working-group.md) suggests that _maintainers_ do it (the "Area label" link referred to below includes `WG-*` labels):

> The issue gets triaged into an [Area label](https://github.com/PowerShell/PowerShell/blob/master/docs/maintainers/issue-management.md#feature-areas) by Maintainers.

What I think is inappropriate is to use deliberate withholding of the `WG*-` as a way of _prejudging_ a feature request (whether or not other WG members jump in and assign the label), and I think it is important to get clarity on that aspect and amend the WG document accordingly.

My reading of the WG document is that `WG-*` labeling should happen _unconditionally_, as the _first step_, as a _formal_ act - and that makes perfect sense to me:

* Once the label is assigned, _all_ members of the relevant WG(s) can weigh in (I presume they have notifications set up for their WG label(s)), ahead of any quorum decision (which may or may not be necessary).

* _Then_ it's perfectly fine to voice opinions on whether a given feature request is "quorum-worthy", but it should never be up to a single WG member to make that call: 

   * If such a pre-quorum conversation is conducted in good faith, including allowing for different opinions, closure may well be achieved for requests if the poster comes to realize themselves that the request isn't worth pursuing.

  * If not, then the opinions already voiced by WG members should aid in making a _quick_ decision in the quorum.

**Comment by mklement0** [comment]
@SteveL-MSFT, I'd like to add the following to my complaint above:

* It is inappropriate for WG members to make unwarranted accusations of trolling in response to carefully crafted arguments proffered in good faith.

* It is inappropriate for WG members to consider themselves the arbiters of who gets to speak when and with how many words (_actual_ violations of the CoC are an entirely unrelated matter).

**Comment by jhoneill** [comment]
Code of conduct talks about 
* Trolling, insulting or derogatory comments,
* harassment
* Disruptive behavior
* conduct which could reasonably be considered inappropriate in a professional setting

Let's just leave repo maintainers to decide what qualifies and what does not.  I feel like I am the subject of vendetta, and they can tell me if I am right or wrong to feel that way. 

WG members do not decide who speaks - that is neither practical nor desirable -but they do decide who the WG listens to and which topics are put before their fellow members. Somewhere I saw "Gatekeeper" used in its pejorative sense. We are portiers,  commissionaires our job is to get people **in**, but we can't do that for everyone. I feel like the nightclub door man who turns one customer away only to be glassed by someone who happened to be passing looking for a fight. In hindsight the initial _" from that perspective nothing that can be implemented in PowerShell code - no matter how cumbersome and non-obvious - constitutes a workaround."_  should have been a warning that the issue was being trolled.  
  
WG membership does not prevent anyone observing that _at some volume_  "carefully crafted arguments" serve no purpose other than to disrupt, and remarking when in their view that volume has been passed.   

If those in charge of the WG think I acted in bad faith they will fire me. The repeated complaint that my actions with tags were not taken in good faith has not resulted in my being told I should have done anything differently, so to the best of my knowledge what I did was legitimate. I have asked the repo maintainers to rule on whether I have good cause to feel harassed  (e.g. the opening of two other issues as thinly veiled, but clearly directed attacks), and if sufficient volume of sufficiently argumentative posts amounts to disruptive behaviour, whether that has been passed. They may say no to both.  As I said already I'm leaving it to them.

**Comment by jhoneill** [comment]
> . It is time to ask those who determine what is and is not trolling to give an opinion.

**Comment by mklement0** [comment]
@jhoneill:

The advantage of conducting continued discussions in the context of an issue that _has already been decided on_ (the feature request was _declined_) is that it can be used to debate things without fear of creating _distractions_ - and, hopefully, useful conclusions will be drawn for _future_ issues.

Yet, your defiant re-posting of part of an earlier comment of yours strikes me as entirely unnecessary.
Please consider hiding or deleting it (and I'll be happy to hide / delete this comment).

**Comment by jhoneill** [comment]
> . I have asked the repo maintainers to rule on whether I have good cause to feel harassed (e.g. the opening of two other issues as thinly veiled, but clearly directed attacks), and if sufficient volume of sufficiently argumentative posts amounts to disruptive behaviour, whether that has been passed. They may say no to both. **As I said already I'm leaving it to them.**

**Comment by mklement0** [comment]
@jhoneill, indeed, you've said that - three times now. Is there a good reason to keep repeating this?

**Comment by doctordns** [comment]
> @jhoneill, indeed, you've said that - three times now. Is there a good reason to keep repeating this?

Is there a good reason for your continued trolling?  

Please stop. If you can’t, then maybe it is time to block you. Your comments in this thread are unhelpful,passive agressive trolling. This really needs to stop.

**Comment by mklement0** [comment]
@doctordns 

Not only was their no _continued_ trolling, there was never any trolling to begin with (more on that below).

I firmly believe my comments in this thread to be helpful to the community at larger, long-term, as they raise important question about tagging, issue triaging, complaint resolution on which community consensus should be reached, and the conclusions should be added to the WG document.

By contrast, unhelpful and passive aggressive does describe your comment and several ones you made in #19730, culminating in [this defiant, argument-free comment](https://github.com/PowerShell/PowerShell/issues/19730#issuecomment-1582845294). But I generally don't think blocking you would be the right answer, and, as argued before, it is _inappropriate for WG members to block community members_, just because these members post comments they dislike and/or misconstrue them as personal attacks. (If members engage in _actual_ CoC violations, then those violations must be dealt with as such.)

@jhoneill's _re-posting_ of his decision not to engage anymore is - to me - inherently unhelpful as well, so I pointed that out _and I offered to clean the unnecessary comments up_. When that offer wasn't taken, I _hid_ my comments, which is the next best thing - you may have noticed that the comments were already hidden at the time you commented.

**Comment by mklement0** [comment]
@SteveL-MSFT

I know this isn't pleasant, and it wasn't my preference to continue this _publicly_ after no shared understanding about WG-member conduct could be reached here, hence my request in #19730.

Since #19730 may take a while to resolve and I knew you had seen this thread, I decided to complain _somewhat_ publicly here, as to me the matter has some urgency. It is unfortunate that #19730 and #19731 are now linked to this issue (one by explicit comments, the other by link only), but that wasn't my choice - in fact, I deliberately wanted to _avoid_ it.

So here we are, with my asking to add another complaint to the list:

* It is inappropriate for WG members to _threaten_ community members with blocking.

As noted, I'm also hoping that the behaviors I have complained about here are added as DO NOTs to the WG document.

**Comment by doctordns** [comment]
Sorry, but you do not have the right to force WG members or anyone else to read posts from trolls.

**Comment by mklement0** [comment]
@SteveL-MSFT, to bring closure to this issue:

* Given the previous frivolous accusations of trolling, @doctordns' [last comment](https://github.com/PowerShell/PowerShell/issues/19680#issuecomment-1584489589) is clearly ad hominem, so I've reported it as abuse to the maintainers, as [now recommended in the WG document](https://github.com/PowerShell/PowerShell/blob/master/docs/community/working-group.md#reporting-working-group-members-abuse).

* @jhoneill has blocked me again. While for users acting as private citizens blocking others is their prerogative, to me it isn't acceptable when acting as a WG member, as [previously argued in more detail](https://github.com/PowerShell/PowerShell/issues/19680#issuecomment-1581260405).

As noted in  #19730, I do hope that additional DON'Ts proscribing the behaviors complained about in this thread are added to the [WG section of the governance document](https://github.com/PowerShell/PowerShell/blob/master/docs/community/governance.md#working-groups-wgs).

**Comment by mklement0** [comment]
One more thing, unfortunately: @doctordns has now inappropriately blocked me too.

**Comment by mklement0** [comment]
@Andrew74L, on a quick meta note: Unless someone formally appeals the WG decision to decline this request (by @-mentioning the committee and making a focused argument), this request is in effect dead (Given what unfolded here, I definitely won't do that.)
That said, continuing the discussion - now without fear of creating a distraction - may still be worthwhile, so that general insights and insights for conceptually related future issues may be gained.

>  perhaps Read-Host should trim by default

I fully agree, and I've mentioned it before, but the reason I didn't suggest it as a solution is that it is technically a _breaking change_; from that perspective, a `-Trim` switch is the _next best thing_.  

That said, at least potentially - I don't have a good sense of whether it would apply here - it could be considered a breaking change that falls into  [bucket 3: Unlikely Grey Area](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md#bucket-3-unlikely-grey-area) which would make it _acceptable_.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee discussed the topic of a number of reported abuses (10 incidents) and whether they broke the Code of Conduct and what appropriate action should be taken.

We do not take the Code of Conduct lightly and letting conversations that break the Conduct may be impacting our ability to attract new community members who only see the negativity that is presumed to be passively allowed.  We want our community and our GitHub repos to be a safe place for professional discourse.

We reviewed each individually reported comment (across different issues) and agreed that the comments violated the Code of Conduct.  Each violator has been blocked from the PowerShell org on GitHub for 30 days.

**Comment by mklement0** [comment]
@SteveL-MSFT, please see https://github.com/PowerShell/PowerShell/issues/19730#issuecomment-1639167491
